noise_corrector.py

```diff
@@ -1,70 +1,177 @@
 #!/usr/bin/env python3
 
 import collections
+import json
+import math
+import numpy
 import random
 
-# This file provides a helper function to correct noisy grous of reports coming
-# from the event level API (i.e. those where conversion metadata is potentially
-# randomly selected). It implements an unbiased estimator for the true counts
-# of conversions for each metadata bucket.
+# This file provides a helper function `estimate_true_values` to correct noisy data coming from the
+# event-level reports in the Attribution Reporting API. It also includes
+# utilities for enumerating the possible outputs from various parametrizations
+# of the event-level reports. See
+# - get_possible_nav_source_outputs
+# - get_possible_event_source_outputs
+# - generate_possible_outputs_per_source
 
-# TODO(csharrison): This estimator is unbiased, but its variance can be
-# improved if we allow bias. This is something we should be able to smoothly
-# trade off with another parameter.
-def corrected_buckets(buckets, noise_probability=.05):
-  """Returns a map of conversion bits --> corrected counts
+# Estimates the true values for data generated through k-ary randomized
+# response. It includes a parameter `beta` which allows smoothly trading off
+# bias (minimized with `beta` = 0) and variance (minimized when `beta` = 1).
+def estimate_true_values(observed_values, randomized_response_rate, beta=0):
+  """Returns a list of corrected counts
 
-  buckets: A map from integer conversion metadata to conversion counts.
-       note, this needs to include buckets with 0 counts.
-  noise_probability: The probability the metadata was randomly selected
+  observed_values: A list of size `k` for k-ary randomized response, where the
+    ith value in the list indicates the counts of the ith output (out of k).
+  randomized_response_rate: The probability any given output is randomized
   """
+  n = sum(observed_values)
+  r = 1. / len(observed_values)
+  x = randomized_response_rate
 
-  total_records = sum(buckets.values())
-  num_conversion_buckets = len(buckets)
+  # An observed value v will be distributed according to
+  # BinomialDistribution[n, q] where
```

Comment:
I don't think this is true. The observed value is distributed according to the convolution of Binom[v', 1 - x * (1 - 1/k)] and Binom[n - v', x / k], where the first distribution corresponds to the counts from users whose true value is v' and the second is the counts from the remaining users.

Reply:
Thank you, I've removed the beta term and just went back to the beta=0 formulation, PTAL